src/flag_gems/ops/weightnorm.py

```diff
@@ -285,5 +374,69 @@ def backward(ctx, w_grad, norm_grad):
         return v_grad, g_grad, None
 
 
+class Norm(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, v, dim):
+        logging.debug("GEMS NORM FORWARD")
+        v = v.contiguous()
+        output = torch.empty(
+            *[1 if i != dim else v.shape[dim] for i in range(v.ndim)],
+            dtype=v.dtype,
+            device=v.device,
+        )
+        v_shape = [
+            math.prod(v.shape[:dim]),
+            v.shape[dim],
+            math.prod(v.shape[dim + 1 :]),
+        ]
+
+        grid = lambda META: (triton.cdiv(v_shape[1], META["BLOCK_ROW_SIZE"]),)
+
+        with torch.cuda.device(v.device):
+            norm_kernel[grid](
+                output,
+                v,
+                v_shape[0],
+                v_shape[1],
+                v_shape[2],
+                eps=torch.finfo(torch.float32).tiny,
+            )
+        ctx.save_for_backward(v, output)
+        ctx.V_SHAPE = v_shape
+        return output
+
+    @staticmethod
+    def backward(ctx, norm_grad):
+        logging.debug("GEMS NORM BACKWARD")
+        norm_grad = norm_grad.contiguous()
+        v, norm = ctx.saved_tensors
+        v_grad = torch.empty_like(v)
+
+        grid = lambda META: (triton.cdiv(ctx.V_SHAPE[1], META["BLOCK_ROW_SIZE"]),)
+        with torch.cuda.device(v.device):
+            norm_bwd_kernel[grid](
+                v_grad,
+                norm_grad,
+                v,
+                norm,
+                ctx.V_SHAPE[0],
+                ctx.V_SHAPE[1],
+                ctx.V_SHAPE[2],
+                eps=torch.finfo(torch.float32).tiny,
+            )
+        return v_grad, None
+
+
+def weight_norm_interface(v, g, dim=0):
+    return WeightNormInterface.apply(v, g, dim)
+
+
 def weight_norm(v, g, dim=0):
-    return WeightNorm.apply(v, g, dim)
+    dim = dim % v.ndim
+    has_half_dtype = v.dtype == torch.float16 or g.dtype == torch.float16
+    can_use_fused = (not has_half_dtype) and (dim == 0 or dim == v.ndim - 1)
+    if can_use_fused:
+        output, _ = weight_norm_interface(v, g, dim)
+        return output
+    else:
+        return v * (g / Norm.apply(v, dim))
```

Comment:
Although Aten also implements the calculation logic through several operators here, can we implement a fused op to speed up it? Or is there a conclusion that fused op cannot accelerate?

Reply:
For cases that don't require broadcasting, the performance of fused operators will be significantly better. However, for cases that do require broadcasting, the performance of fused operators is similar to the current implementation (since data broadcasting involves actual copying). How about I add a new implementation specifically for cases without broadcasting, while keeping the current approach for more complex broadcasting cases?

Reply:
> For cases that don't require broadcasting, the performance of fused operators will be significantly better. However, for cases that do require broadcasting, the performance of fused operators is similar to the current implementation (since data broadcasting involves actual copying). How about I add a new implementation specifically for cases without broadcasting, while keeping the current approach for more complex broadcasting cases?

I am also not sure whether fusion operators can bring significant performance gains. So we can fuse it later, and merge this version now.